### DISCUSSION TITLE
Why do DTW distance metrics behave differently?

### DISCUSSION THREAD

**Discussion by fergus-pick** [alan-turing-institute/sktime#5814] [Q&A] [answered]
I am new to using sktime, I was wondering why the `numba` `sktime.distances` `metric='dtw'` gives a different output than the `sktime.dists_kernels` implementation for this case, yet the squared metric is identical?

```python
from sktime.datasets import load_italy_power_demand
from sktime.dists_kernels import DtwDist
from sktime.distances import distance

X, _ = load_italy_power_demand(split="train", return_type="numpy3D")
dtw_dist = DtwDist()
dtw_dist(X[0], X[1])
distance(X[0], X[1], metric='dtw')
distance(X[0], X[1], metric='squared')

>>> [[4.88675167]]
>>> 2.2562779371935555
>>> 4.886751673604364

```

**Comment by fkiraly** [comment, marked as answer]
Update: I know what is going on, this is not a bug.

`dtw_dist` is a pairwise distance, so expects - consistently with all other panel data estimators - input in one of the `sktime` `Panel` formats, e.g., `numpy3D` with `(n_instance, n_variable, n_timepoints)`.

So, if you pass `dtw_dist(X, X)`, you will get a matrix where the (0,1)-the and (1,0)-th element are, as expected, 2.256etc.

I am not sure why it runs with 2D `numpy`, it seems to assume a single instance with 24 variables and 1 time point. For which the DTW distance is the same as the squared distance, if the inner distance (on scalars) is squared euclidean.

That would explain the behaviour?

----
Old post (incorrect)

Hm, that is curious. `DtwDist` should be a direct interface to `distance`.

However, upon closer inspection, `DtwDist` interfaces `pairwise_distance`, which internally has different code to `distance`, but both claim to be the same (up to a loop).
My guess is that one of the two is buggy, this seems like a mixup between numpy 2D, numpy 3D, index order or formatting or the like.

Do you perchance know which one is correct? We could calculate `"squared"` maually.
Also, the internal relevant code from `pairwise_distance` vs `distance`, below:

`pairwise_distance`:

```python
    from sktime.distances._numba_utils import (
        _compute_pairwise_distance,
        _make_3d_series,
    )

    _x = _make_3d_series(x)
    if y is None:
        y = x
    _y = _make_3d_series(y)
    symmetric = np.array_equal(_x, _y)
    _metric_callable = _resolve_metric_to_factory(
        metric, _x[0], _y[0], _METRIC_INFOS, **kwargs
    )
```

`distance`:
```python
    from sktime.distances._numba_utils import to_numba_timeseries

    _x = to_numba_timeseries(x)
    _y = to_numba_timeseries(y)

    _metric_callable = _resolve_metric_to_factory(
        metric, _x, _y, _METRIC_INFOS, **kwargs
    )

```

**Reply by fergus-pick** [reply]
Thanks for your quick reply, that explains the behavior. I have now amended my sample code to show the expected output by adding the `n_instance` dimension.
```python
X, _ = load_italy_power_demand(split="train", return_type="numpy3D")
dtw_dist = DtwDist()
dtw_dist(np.expand_dims(X[0], 0), np.expand_dims(X[1], 0))
distance(X[0], X[1], metric='dtw')
distance(X[0], X[1], metric='squared')
>>> [[2.25627794]]
>>> 2.2562779371935555
>>> 4.886751673604364
```

If when passing in 2D `numpy` format with shape `(n_variable, n_timepoints)`, it assumes the sample as `(1, n_timepoints, n_variables)`, then is the behavior of the below code unusual as the 2D `numpy` `dtw_dist` output does not match the squared distance? When I reshape to `(1, n_timepoints, n_variables)`, the result from the 2D `dtw_dist` matches the squared distance, so I wonder what is happening in the first`dtw_dist` call?

```python
from sktime.datasets import load_basic_motions
from sktime.dists_kernels import DtwDist
from sktime.distances import distance

X, _ = load_basic_motions(split="train", return_type="numpy3D")
print(X[0].shape)
dtw_dist = DtwDist()
print(dtw_dist(X[0], X[1]))
print(distance(X[0], X[1], metric='dtw'))
print(distance(X[0], X[1], metric='squared'))

reshape_X_0 = np.reshape(X[0], (1, 100, 6)) # (1 instance, n_timepoints, n_variables)
reshape_X_1 = np.reshape(X[1], (1, 100, 6))
print(dtw_dist(reshape_X_0, reshape_X_1))

>>> (6, 100) # sample shape
>>> [[581.78520127]] # 2D DtwDist()
>>> 330.8344972144629 # sktime.distances 'dtw'
>>> 757.25971908652 # sktime.distances 'squared'
>>> [[757.25971909]] # 3D DtwDist() after reshaping
```

**Reply by fkiraly** [reply]
At the bottom are the mtype (datatype) specifications for `Panel` data that `sktime` recognizes,

I think internally it recognizes the type as `numpyflat` and converts it to `numpy3D`.
However I'm confused then why the result is obtained.

The conversion code (relevant lines only) is:

```python
    target_shape = (shape[0], 1, shape[1])
    obj_in_3D = obj.reshape(target_shape)
```

List of mtypes:

```
MTYPE_REGISTER_PANEL = [
    (
        "nested_univ",
        "Panel",
        "pd.DataFrame with one column per variable, pd.Series in cells",
    ),
    (
        "numpy3D",
        "Panel",
        "3D np.array of format (n_instances, n_columns, n_timepoints)",
    ),
    (
        "numpyflat",
        "Panel",
        "WARNING: only for internal use, not a fully supported Panel mtype. "
        "2D np.array of format (n_instances, n_columns*n_timepoints)",
    ),
    ("pd-multiindex", "Panel", "pd.DataFrame with multi-index (instances, timepoints)"),
    ("pd-wide", "Panel", "pd.DataFrame in wide format, cols = (instance*timepoints)"),
    (
        "pd-long",
        "Panel",
        "pd.DataFrame in long format, cols = (index, time_index, column)",
    ),
    ("df-list", "Panel", "list of pd.DataFrame"),
    (
        "dask_panel",
        "Panel",
        "dask frame with one instance and one time index, as per dask_to_pd convention",
    ),
]
```

**Reply by fkiraly** [reply]
In case you wonder why there are so many, this is a collection of representations different developers within `sktime` and adjacent packages have been using to represent "collections of time series" aka `Panel` abstract data type.

To integrate all under a unified interface without having to rewrite dozens of different locations, we went with a converter/representation architecture hidden in the boilerplate layer.

**Reply by fkiraly** [reply]
> then is the behavior of the below code unusual 

I think your assumed specification is not what it is. According to the specification sheet:

* 3D numpy is `(n_instances, n_variables, n_timepoints)` (not `n_instances, n_timepoints, n_variables)` as you assume)
* 2D numpy is `(n_instances, n_timepoints)` (not `(n_variables, n_timepoints)` as you assume)

Somewhat unfortunantely, but I think there is no way around it, one interpretation of 3D and 2D numpy has to be enforced across all of `sktime`. There has been a lot of discussion of this in early days, but one can't interpret 3D numpy on way in one module and another way in another, same for 2D numpy.

**Reply by fkiraly** [reply]
tbh, I think the `pd.DataFrame` based representation is less confusing...

**Reply by fergus-pick** [reply]
Thanks for spending the time to explain this, it makes more sense now.

**Reply by fkiraly** [reply]
if you have any suggestions on how to improve either documentation of interface, these would be very welcome - especially if critical
